dvc/ignore.py

```diff
@@ -71,10 +71,25 @@ def __eq__(self, other):
         return self.basenames == other.basenames
 
 
+class DvcIgnoreRepo(DvcIgnore):
+    def __call__(self, root, dirs, files):
+        def is_dvc_repo(directory):
+            from dvc.repo import Repo
+
+            return os.path.isdir(os.path.join(directory, Repo.DVC_DIR))
+
+        dirs = [d for d in dirs if not is_dvc_repo(d)]
+
+        return dirs, files
+
+
 class DvcIgnoreFilter(object):
     def __init__(self, tree):
         self.tree = tree
-        self.ignores = {DvcIgnoreDirs([".git", ".hg", ".dvc"])}
+        self.ignores = {
+            DvcIgnoreDirs([".git", ".hg", ".dvc"]),
```

Comment:
.hg ? Do we support Mercurial somehow now??? Or is this legacy stuff? Thanks

Reply:
We don't want to traverse through `.git` or `.hg` folders. People might use `hg` for code versioning and use `--no-scm` with DVC.

Reply:
Hm... Why only do this for .hg? There's many many possible files or dirs the user may want to gitignore and they can just add it to .gitignore/ or .git/info/exclude

Reply:
@jorgeorpinel add `.hg` to `.gitignore`? Maybe I'm missing something.

An alternative would be to add `.hg` to `.dvcignore` file, yes, but since we know about `.hg`, we are skipping it automatically.

Reply:
> add .hg to .gitignore?

Haha sorry I wasn't thinking right 😋 

> to add .hg to .dvcignore

Yeah, this! Since we don't support Mercurial in general why treat it specially? I'd let the user handle it, like he could want dvcignore any other files or directories for a bunch of reasons.

Reply:
> Since we don't support Mercurial in general why treat it specially?

Why traverse through such directories at all? It'll only make dvc slower. It's a reasonable default.

Reply:
> such directories 

Exactly. There can be tons of special/hidden dirs you don't want DVC to traverse (.idea/ .project/ .env/ .ds_store/ venv/), and you can just add them to .gitignore or .dvcignore. Why are we secretly integrating with Mercurial if we don't actually support it? What if someone wants to version a .hg dir? (edge case I know)

It just seems very arbitrary to me, but OK maybe I'm missing something...

Reply:
My 2cs :) Not important thing, but I also don't understand why do we ignore some `.hg` dir which does not mean anything to us. It is some magic behavior not related to the product features in any way. The good example would be - does Git ignore .hg by default?

Reply:
> does Git ignore .hg by default?

No, it doesn't. And, why should it? It's not meant to work with other version controls. whereas, dvc needs to work with `git` and any other version controls out there. Other stuffs like `__pycache__`, `venv` can be assumed to be ignored in `.gitignore` or `.dvcignore`.
`.hg` is really a special case.

Do people really use `.hg` folders for their work? Unless they do (which they shouldn't), benefits outweigh the drawbacks.

Reply:
@skshetry 

>  and any other version controls out there

that would be true if DVC worked with hg, right? It does not right now, and we don't have any plans do this.

> Do people really use .hg folders for their work? Unless they do (which they shouldn't), benefits outweigh the drawbacks.

it's a slippery slope. This logic applies easily to all other aforementioned cases like DSStore, etc. That's why there is a mechanism (and no list of defaults in Git for this).

Another example (if you don't like git itself) is `git-lfs`. DVC with its data management layer is kind-a the same extension to Git as git-lfs. I doubt that git-lfs, or git-annex introduce a list of defaults on top.

Reply:
I just don't really see any difference between __pycache__, venv and .hg, really. I would say even including __pycache__ would even more useful by default.

.git and .dvc feel like the only special cases for DVC right now.

Reply:
> git-lfs

It only works on top of git. Not alongside `git` as DVC does. So, it makes sense there.

DVC has `--no-scm`, so they could be using any version control out there or none. The other one except of `git` that's popular is `hg`, so, I think it makes sense to ignore by default.

And, how many people are using or are aware of `.dvcignore`?

Reply:
> And, how many people are using or are aware of .dvcignore?

yep, this applies to all other example mentioned above we should ignore then? A lot of them make even more value.

> ... that's popular is hg

perforce and svn have a bigger market share? I doubt that mercurial has more than a few %% (which is sad).

--------------

Interesting discussion. I realized that there might be indeed a stronger motivation to include more defaults (`.hg` is still strange and not the most painful example) into DVC ignore list. The difference between them is that for DVC it **affects performance silently** (imagine analyzing `.node_modules`), for Git it affects `git status` and **you immediately see that** this is wrong, and start looking for `.gitingnore`, etc. You won't have a motivation do this with DVC - I mean motivation to start looking for `.dvcignore` unless you need to actually ignore some files doing `dvc add`?

Does it make sense to include all files that are in `.gitignore` into `.dvcignore` ? (may be we already do this?) Does it makes sense to actually provide a bigger list of defaults?

Reply:
> DVC has --no-scm, so they could be using any version control out there

I don't think this is the case @skshetry. It's just kind of a legacy option name AFAIK. Refer to #2901. 

> Do people really use .hg folders for their work? Unless they do (which they shouldn't)

You'd be surprised. Why shouldn't they? Will the directory name police pul them over? 😛 

> I realized that there might be indeed a stronger motivation to include more defaults

@shcheklein I don't even see this. DVC ignores .git/ automatically because it integrates with Git. I don't see a reason to treat any other file/dir name specially. Your questions about .gitignore vs. .dvcignore are interesting though – I think there may be some issues already about that.

---

In summary, I vote to remove the 4 characters we're talking about, but I realize this is probably a moot point since no one even uses Mercurial anymore haha.

Reply:
@jorgeorpinel, I think something good thing to do is in middle. If people have a certain directories such as `__pycache__` and `venv` or similar kinds of directories, it'd be good to make them aware that `.dvcignore` exists, at least a warning that traversing will be slow.

Reply:
Agree about improving /dvcignore awareness. (I think the warning may exist BTW?)

For docs, I opened iterative/dvc.org/issues/1033.

Reply:
> at least a warning that traversing will be slow.

it's a really good idea. Please create an issue for this in the DVC core repo.

I still don't see a reason to include `.hg` though. To my mind if we follow this logic it should be done properly like you described for example.

Reply:
Created #3436

Reply:
Resolving this for now.